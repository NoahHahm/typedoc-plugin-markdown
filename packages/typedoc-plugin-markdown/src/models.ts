import { ReflectionKind, TypeDocOptionMap } from 'typedoc';

export interface TypedocPluginMarkdownOptions extends TypeDocOptionMap {
  anchorFormat: AnchorFormat;
  baseUrl: string;
  enableFrontmatter: boolean;
  entryDocument: string;
  enumMembersFormat: DataFormatStyle;
  flattenOutputFiles: boolean;
  frontmatterTags: string[];
  frontmatterGlobals: FrontmatterGlobals;
  frontmatterNamingConvention: FrontmatterNamingConvention;
  groupByKinds: boolean;
  hideBreadcrumbs: boolean;
  hideInPageTOC: boolean;
  hideKindTag: boolean;
  hidePageTitle: boolean;
  hideHierarchy: boolean;
  indentifiersAsCodeBlocks: boolean;
  indexPageTitle: string;
  kindsWithOwnFile: string | string[];
  namedAnchors: boolean;
  numberPrefixOutput: boolean;
  propertiesFormat: DataFormatStyle;
  TOCFormat: DataFormatStyle;
}

export interface TemplateMapping {
  directory: string | null;
  template: any;
  isLeaf: boolean;
  kind: ReflectionKind;
}

export type Collapse = 'object' | 'function' | 'all' | 'none';

export type DataFormatStyle = 'list' | 'table';

export type AnchorFormat = 'lowercase' | 'slug' | 'none';

export type FrontmatterGlobals =
  | string
  | Record<string, string | boolean | number | null>;

export type FrontmatterNamingConvention =
  | 'camelCase'
  | 'snakeCase'
  | 'kebabCase'
  | 'pascalCase';

export interface NavigationItem {
  title?: string;
  url?: string;
  children?: NavigationItem[];
}
