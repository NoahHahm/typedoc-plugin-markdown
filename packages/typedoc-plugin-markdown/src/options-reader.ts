import { Options, OptionsReader } from 'typedoc';
import { TypedocPluginMarkdownOptions } from './models';

export class MarkdownPluginOptionsReader implements OptionsReader {
  priority = 900;
  options: Partial<TypedocPluginMarkdownOptions>;

  name = 'custom-options';

  constructor(options: Partial<TypedocPluginMarkdownOptions>) {
    this.options = options;
  }

  read(container: Options) {
    if (this.options) {
      Object.entries(this.options).forEach(([key, value]) => {
        container.setValue(key, value);
      });
    }
  }
}
