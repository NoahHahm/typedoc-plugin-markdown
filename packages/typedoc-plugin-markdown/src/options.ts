import { Application, Options, OptionsReader, ParameterType } from 'typedoc';

export function defineOptions(app: Application) {
  app.options.addReader(
    new (class implements OptionsReader {
      name = 'markdown-theme-options-reader';
      readonly order = 900;
      readonly supportsPackages = false;
      read(container: Options) {
        if (container.getValue('theme') === 'default') {
          container.setValue('theme', 'markdown');
        }
      }
    })(),
  );

  /**
   * file output options
   */
  app.options.addDeclaration({
    name: 'entryDocument',
    help: '[Markdown Plugin] The file name of the entry document.',
    type: ParameterType.String,
    defaultValue: 'README.md',
  });

  app.options.addDeclaration({
    name: 'flattenOutput',
    help: '[Markdown Plugin] Flatten output files without directories.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'numberPrefixOutput',
    help: '[Markdown Plugin] Prefixes docs and folders by number prefixes if applicable.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'kindsWithOwnFile',
    help: "[Markdown Plugin] Specifies which reflection kinds are contained in their own file. Values 'none', 'all' OR Array of ['class', 'interface', 'enum', 'function', 'variable', 'type']",
    type: ParameterType.String | ParameterType.Array,
    defaultValue: 'all',
  });

  app.options.addDeclaration({
    name: 'groupByKinds',
    help: '[Markdown Plugin] Groups reflection kinds by headings if applicable e.g Classes, Functions. If set to false all symbols will render on the same level. Defaults to `true`',
    type: ParameterType.Boolean,
    defaultValue: true,
  });

  /**
   * ui options
   */

  app.options.addDeclaration({
    name: 'hidePageTitle',
    help: '[Markdown Plugin] Do not print page title.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'hideKindTag',
    help: '[Markdown Plugin] Do not print the kind tag beneath the page title.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'hideBreadcrumbs',
    help: '[Markdown Plugin] Do not print breadcrumbs.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'hideInPageTOC',
    help: '[Markdown Plugin] Do not render in-page table of contents items.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'hideHierarchy',
    help: '[Markdown Plugin] Do not print reflection hierarchy.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'indexPageTitle',
    help: '[Markdown Plugin] The title of the main index / modules page.',
    type: ParameterType.String,
  });

  app.options.addDeclaration({
    name: 'indentifiersAsCodeBlocks',
    help: '[Markdown Plugin] Format signature and declaration identifiers in code blocks.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'propertiesFormat',
    help: '[Markdown Plugin] Specify the render style of properties.',
    type: ParameterType.String,
    defaultValue: 'List',
  });

  app.options.addDeclaration({
    name: 'enumMembersFormat',
    help: '[Markdown Plugin] Specify the render style of properties.',
    type: ParameterType.String,
    defaultValue: 'List',
  });

  app.options.addDeclaration({
    name: 'typeDeclarationFormat',
    help: '[Markdown Plugin] Specify the render style of type declarations.',
    type: ParameterType.String,
    defaultValue: 'List',
  });

  /**
   * Frontmatter options
   */
  app.options.addDeclaration({
    name: 'enableFrontmatter',
    help: '[Markdown Plugin] Prepend frontmatter to output.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'frontmatterTags',
    help: '[Markdown Plugin] Specify which file comment tags should be added to frontmatter.',
    type: ParameterType.Array,
    defaultValue: [],
  });

  app.options.addDeclaration({
    name: 'frontmatterGlobals',
    help: '[Markdown Plugin] Specify static variables to be added to all frontmatter.',
    type: ParameterType.Mixed,
    defaultValue: {},
  });

  app.options.addDeclaration({
    name: 'frontmatterNamingConvention',
    help: '[Markdown Plugin] Specify the naming convention of front matter variables.',
    type: ParameterType.String,
    defaultValue: 'camelCase',
    validate: (option) => {
      const availableValues = [
        'camelCase',
        'snakeCase',
        'kebabCase',
        'pascalCase',
      ];
      if (!availableValues.includes(option)) {
        throw new Error(
          `Unexpected value for frontmatterNamingConvention, the expected value is one of 'camelCase', 'snakeCase','kebabCase','pascalCase'`,
        );
      }
    },
  });

  /**
   * Utility options
   */
  app.options.addDeclaration({
    help: '[Markdown Plugin] Specifies the base url for internal link. If omitted all urls will be relative.',
    name: 'baseUrl',
    type: ParameterType.String,
  });

  app.options.addDeclaration({
    name: 'namedAnchors',
    help: '[Markdown Plugin] Use HTML named anchors as fragment identifiers for engines that do not automatically assign header ids. Should be set for Bitbucket Server docs.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'preserveAnchorCasing',
    help: '[Markdown Plugin] Preserve anchor casing when generating links.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });
}
