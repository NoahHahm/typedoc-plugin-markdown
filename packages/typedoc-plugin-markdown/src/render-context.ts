import * as path from 'path';
import { Options, PageEvent, Reflection } from 'typedoc';
import { URL_PREFIX } from './support/constants';

/* start_imports */
import { breadcrumbs } from './resources/partials/breadcrumbs';
import { commentParts } from './resources/partials/comment.parts';
import { comment } from './resources/partials/comment';
import { footer } from './resources/partials/footer';
import { header } from './resources/partials/header';
import { hierarchy } from './resources/partials/hierarchy';
import { declarationMemberIdentifier } from './resources/partials/member.declaration.identifier';
import { declarationMemberName } from './resources/partials/member.declaration.name';
import { declarationMember } from './resources/partials/member.declaration';
import { memberIndex } from './resources/partials/member.index';
import { indexSignatureTitle } from './resources/partials/member.indexsignature.title';
import { inheritance } from './resources/partials/member.inheritance';
import { memberKindTag } from './resources/partials/member.kind-tag';
import { referenceMember } from './resources/partials/member.reference';
import { reflectionMember } from './resources/partials/member.reflection';
import { signatureMemberIdentifier } from './resources/partials/member.signature.identifier';
import { signatureMember } from './resources/partials/member.signature';
import { sources } from './resources/partials/member.sources';
import { memberTitle } from './resources/partials/member.title';
import { member } from './resources/partials/member';
import { typeDeclarationMember } from './resources/partials/member.type-declaration';
import { members } from './resources/partials/members';
import { navigation } from './resources/partials/navigation';
import { pageIndex } from './resources/partials/page.index';
import { pageTitle } from './resources/partials/page.title';
import { enumMembersTable } from './resources/partials/table.enum-members';
import { parametersTable } from './resources/partials/table.parameters';
import { propertiesTable } from './resources/partials/table.properties';
import { typeParametersTable } from './resources/partials/table.typeparameters';
import { arrayType } from './resources/partials/type.array';
import { conditionalType } from './resources/partials/type.conditional';
import { declarationType } from './resources/partials/type.declaration';
import { functionType } from './resources/partials/type.function';
import { indexAccessType } from './resources/partials/type.index-access';
import { inferredType } from './resources/partials/type.inferred';
import { intersectionType } from './resources/partials/type.intersection';
import { intrinsicType } from './resources/partials/type.intrinsic';
import { literalType } from './resources/partials/type.literal';
import { queryType } from './resources/partials/type.query';
import { referenceType } from './resources/partials/type.reference';
import { reflectionType } from './resources/partials/type.reflection';
import { someType } from './resources/partials/type.some';
import { tupleType } from './resources/partials/type.tuple';
import { typeOperatorType } from './resources/partials/type.type-operator';
import { unionType } from './resources/partials/type.union';
import { unknownType } from './resources/partials/type.unknown';
import { memberTemplate } from './resources/templates/member';
import { projectTemplate } from './resources/templates/project';
import { readmeTemplate } from './resources/templates/read-me';
import { reflectionTemplate } from './resources/templates/reflection';
/* end_imports */

function bind<F, L extends any[], R>(fn: (f: F, ...a: L) => R, first: F) {
  return (...r: L) => fn(first, ...r);
}

export class MarkdownThemeRenderContext {
  constructor(
    public page: PageEvent<Reflection> | null,
    public options: Options,
  ) {}

  urlTo = (reflection: Reflection) => {
    return this.relativeURL(reflection.url);
  };

  relativeURL = (url: string | undefined) => {
    if (!url) {
      return null;
    }
    if (URL_PREFIX.test(url)) {
      return url;
    } else {
      if (this.options.getValue('baseUrl')) {
        return this.options.getValue('baseUrl') + url.replace(/\\/g, '/');
      }

      const relative = path.relative(
        path.dirname(this.page?.url || '.'),
        path.dirname(url),
      );

      return this.parseUrl(
        path.join(relative, path.basename(url)).replace(/\\/g, '/'),
      );
    }
  };

  parseUrl(url: string) {
    return encodeURI(url);
  }

  /* start_resources */

  // templates
  memberTemplate = bind(memberTemplate, this);
  projectTemplate = bind(projectTemplate, this);
  readmeTemplate = bind(readmeTemplate, this);
  reflectionTemplate = bind(reflectionTemplate, this);

  // partials
  breadcrumbs = bind(breadcrumbs, this);
  commentParts = bind(commentParts, this);
  comment = bind(comment, this);
  footer = bind(footer, this);
  header = bind(header, this);
  hierarchy = bind(hierarchy, this);
  declarationMemberIdentifier = bind(declarationMemberIdentifier, this);
  declarationMemberName = bind(declarationMemberName, this);
  declarationMember = bind(declarationMember, this);
  memberIndex = bind(memberIndex, this);
  indexSignatureTitle = bind(indexSignatureTitle, this);
  inheritance = bind(inheritance, this);
  memberKindTag = bind(memberKindTag, this);
  referenceMember = bind(referenceMember, this);
  reflectionMember = bind(reflectionMember, this);
  signatureMemberIdentifier = bind(signatureMemberIdentifier, this);
  signatureMember = bind(signatureMember, this);
  sources = bind(sources, this);
  memberTitle = bind(memberTitle, this);
  member = bind(member, this);
  typeDeclarationMember = bind(typeDeclarationMember, this);
  members = bind(members, this);
  navigation = bind(navigation, this);
  pageIndex = bind(pageIndex, this);
  pageTitle = bind(pageTitle, this);
  enumMembersTable = bind(enumMembersTable, this);
  parametersTable = bind(parametersTable, this);
  propertiesTable = bind(propertiesTable, this);
  typeParametersTable = bind(typeParametersTable, this);
  arrayType = bind(arrayType, this);
  conditionalType = bind(conditionalType, this);
  declarationType = bind(declarationType, this);
  functionType = bind(functionType, this);
  indexAccessType = bind(indexAccessType, this);
  inferredType = bind(inferredType, this);
  intersectionType = bind(intersectionType, this);
  intrinsicType = bind(intrinsicType, this);
  literalType = bind(literalType, this);
  queryType = bind(queryType, this);
  referenceType = bind(referenceType, this);
  reflectionType = bind(reflectionType, this);
  someType = bind(someType, this);
  tupleType = bind(tupleType, this);
  typeOperatorType = bind(typeOperatorType, this);
  unionType = bind(unionType, this);
  unknownType = bind(unknownType, this);
  /* end_resources */
}
