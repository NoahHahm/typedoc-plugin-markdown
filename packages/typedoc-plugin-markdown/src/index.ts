import { Application } from 'typedoc';
import { defineOptions } from './options';
import { generateMarkdown, renderMarkdown } from './renderer';
import { MarkdownTheme } from './theme';
import { MarkdownThemeRenderContext } from './theme-context';

export function load(app: Application) {
  /**
   * Exposes markdown theme to the renderer
   */
  app.renderer.defineTheme('markdown', MarkdownTheme);

  /**
   * Defines all plugin options
   */
  defineOptions(app);

  /**
   * Decouple HTML logic from the renderer (there should probably be a better solution to this)
   */
  Object.defineProperty(app, 'generateDocs', { value: generateMarkdown });
  Object.defineProperty(app.renderer, 'render', {
    value: renderMarkdown,
    configurable: true,
  });
}

/**
 * Expose global entrypoints
 */
export * from './models';
export * from './options-reader';
export { partials } from './resources';
export { MarkdownTheme, MarkdownThemeRenderContext };
