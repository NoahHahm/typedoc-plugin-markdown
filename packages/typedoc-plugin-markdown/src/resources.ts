// THIS FILE IS AUTOGENERATED - DO NOT EDIT DIRECTLY
import { 
 ArrayType,
 ConditionalType,
 ContainerReflection,
 Comment,
 CommentDisplayPart,
 DeclarationHierarchy,
 DeclarationReflection,
 InferredType,
 IntersectionType,
 IntrinsicType,
 IndexedAccessType,
 LiteralType,
 PageEvent,
 ParameterReflection,
 ProjectReflection,
 QueryType,
 ReferenceReflection,
 ReferenceType,
 ReflectionType,
 SignatureReflection,
 SomeType,
 TupleType,
 TypeOperatorType,
 TypeParameterReflection,
 UnionType,
 UnknownType } from 'typedoc';
import { MarkdownThemeRenderContext } from './theme-context';
import { Collapse } from './models';
import { memberTemplate } from './templates/member';
import { projectTemplate } from './templates/project';
import { readmeTemplate } from './templates/readme';
import { reflectionTemplate } from './templates/reflection';
import { breadcrumbs } from './partials/breadcrumbs';
import { commentParts } from './partials/comment.parts';
import { comment } from './partials/comment';
import { header } from './partials/header';
import { hierarchy } from './partials/hierarchy';
import { declarationMemberTitle } from './partials/member.declaration.title';
import { declarationMember } from './partials/member.declaration';
import { indexSignatureTitle } from './partials/member.indexsignature.title';
import { referenceMember } from './partials/member.reference';
import { signatureTitle } from './partials/member.signature.title';
import { signatureMember } from './partials/member.signature';
import { member } from './partials/member';
import { typeDeclarationList } from './partials/member.typedeclaration.list';
import { typeDeclarationTable } from './partials/member.typedeclaration.table';
import { typeParameters } from './partials/member.typeparameters';
import { members } from './partials/members';
import { reflectionTitle } from './partials/reflection.title';
import { reflection } from './partials/reflection';
import { sources } from './partials/sources';
import { parametersTable } from './partials/table.parameters';
import { toc } from './partials/toc';
import { arrayType } from './partials/type.array';
import { conditionalType } from './partials/type.conditional';
import { declarationType } from './partials/type.declaration';
import { functionType } from './partials/type.function';
import { indexAccessType } from './partials/type.indexAccess';
import { inferredType } from './partials/type.inferred';
import { intersectionType } from './partials/type.intersection';
import { intrinsicType } from './partials/type.intrinsic';
import { literalType } from './partials/type.literal';
import { queryType } from './partials/type.query';
import { referenceType } from './partials/type.reference';
import { reflectionType } from './partials/type.reflection';
import { someType } from './partials/type.some';
import { tupleType } from './partials/type.tuple';
import { typeOperatorType } from './partials/type.typeoperator';
import { unionType } from './partials/type.union';
import { unknownType } from './partials/type.unknown';

function bind<F, L extends any[], R>(fn: (f: F, ...a: L) => R, first: F) {
  return (...r: L) => fn(first, ...r);
}
export type Templates = {
  memberTemplate: (page: PageEvent<DeclarationReflection>) => string;
  projectTemplate: (page: PageEvent<ProjectReflection>) => string;
  readmeTemplate: (page: PageEvent<ProjectReflection>) => string;
  reflectionTemplate: (page: PageEvent<DeclarationReflection>) => string;
};

export type Partials = {
  breadcrumbs: (page: PageEvent<DeclarationReflection | ProjectReflection>) => string;
  commentParts: (parts: CommentDisplayPart[]) => string;
  comment: (comment: Comment) => string;
  header: (page: PageEvent<DeclarationReflection>) => string;
  hierarchy: (declarationHierarchy: DeclarationHierarchy) => string;
  declarationMemberTitle: (reflection: DeclarationReflection | ParameterReflection) => string;
  declarationMember: (declaration: DeclarationReflection) => string;
  indexSignatureTitle: (signature: SignatureReflection) => string;
  referenceMember: (props: ReferenceReflection) => string;
  signatureTitle: (signature: SignatureReflection, accessor?: string |  undefined) => string;
  signatureMember: (signature: SignatureReflection, parentHeadingLevel?: number |  undefined) => string;
  member: (reflection: DeclarationReflection) => string;
  typeDeclarationList: (props: DeclarationReflection[]) => string;
  typeDeclarationTable: (props: DeclarationReflection[]) => string;
  typeParameters: (typeParameters: TypeParameterReflection[]) => string;
  members: (container: ContainerReflection) => string;
  reflectionTitle: (reflection: DeclarationReflection) => string;
  reflection: (reflection: DeclarationReflection) => string;
  sources: (reflection: DeclarationReflection | SignatureReflection) => string;
  parametersTable: (parameters: ParameterReflection[]) => string;
  toc: (reflection: DeclarationReflection | ProjectReflection) => string;
  arrayType: (arrayType: ArrayType) => string;
  conditionalType: (conditionalType: ConditionalType) => string;
  declarationType: (declarationReflection: DeclarationReflection, collapse?: Collapse) => string;
  functionType: (modelSignatures: SignatureReflection[]) => string;
  indexAccessType: (model: IndexedAccessType) => string;
  inferredType: (model: InferredType) => string;
  intersectionType: (model: IntersectionType) => string;
  intrinsicType: (model: IntrinsicType) => string;
  literalType: (literalType: LiteralType) => string;
  queryType: (queryType: QueryType) => string;
  referenceType: (referenceType: ReferenceType) => string;
  reflectionType: (reflectionType: ReflectionType, collapse: Collapse) => string;
  someType: (someType: SomeType, collapse?: Collapse) => string;
  tupleType: (tupleType: TupleType) => string;
  typeOperatorType: (model: TypeOperatorType) => string;
  unionType: (unionType: UnionType) => string;
  unknownType: (model: UnknownType) => string;
};

export const templates = (context: MarkdownThemeRenderContext): Templates => ({
  memberTemplate: bind(memberTemplate, context),
  projectTemplate: bind(projectTemplate, context),
  readmeTemplate: bind(readmeTemplate, context),
  reflectionTemplate: bind(reflectionTemplate, context),
});

export const partials = (context: MarkdownThemeRenderContext): Partials => ({
  breadcrumbs: bind(breadcrumbs, context),
  commentParts: bind(commentParts, context),
  comment: bind(comment, context),
  header: bind(header, context),
  hierarchy: bind(hierarchy, context),
  declarationMemberTitle: bind(declarationMemberTitle, context),
  declarationMember: bind(declarationMember, context),
  indexSignatureTitle: bind(indexSignatureTitle, context),
  referenceMember: bind(referenceMember, context),
  signatureTitle: bind(signatureTitle, context),
  signatureMember: bind(signatureMember, context),
  member: bind(member, context),
  typeDeclarationList: bind(typeDeclarationList, context),
  typeDeclarationTable: bind(typeDeclarationTable, context),
  typeParameters: bind(typeParameters, context),
  members: bind(members, context),
  reflectionTitle: bind(reflectionTitle, context),
  reflection: bind(reflection, context),
  sources: bind(sources, context),
  parametersTable: bind(parametersTable, context),
  toc: bind(toc, context),
  arrayType: bind(arrayType, context),
  conditionalType: bind(conditionalType, context),
  declarationType: bind(declarationType, context),
  functionType: bind(functionType, context),
  indexAccessType: bind(indexAccessType, context),
  inferredType: bind(inferredType, context),
  intersectionType: bind(intersectionType, context),
  intrinsicType: bind(intrinsicType, context),
  literalType: bind(literalType, context),
  queryType: bind(queryType, context),
  referenceType: bind(referenceType, context),
  reflectionType: bind(reflectionType, context),
  someType: bind(someType, context),
  tupleType: bind(tupleType, context),
  typeOperatorType: bind(typeOperatorType, context),
  unionType: bind(unionType, context),
  unknownType: bind(unknownType, context),
});
