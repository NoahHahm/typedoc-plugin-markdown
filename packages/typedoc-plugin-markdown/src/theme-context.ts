import * as path from 'path';
import { DeclarationReflection, Options, PageEvent, Reflection } from 'typedoc';
import { TypedocPluginMarkdownOptions } from './models';
import { partials, templates } from './resources';
import { URL_PREFIX } from './support/constants';
import { getReflectionTitle } from './support/helpers';
import { unEscapeChars } from './support/utils';
import { MarkdownTheme } from './theme';

export class MarkdownThemeRenderContext {
  public globalsFile = 'modules.md';

  private _activeLocation: string;

  set activeLocation(activeLocation: string) {
    this._activeLocation = activeLocation;
  }

  get activeLocation() {
    return this._activeLocation;
  }

  constructor(public theme: MarkdownTheme, public options: Options) {}

  getOption<K extends keyof TypedocPluginMarkdownOptions>(name: K) {
    return this.options.getValue(name) as TypedocPluginMarkdownOptions[K];
  }

  getRelativeUrl(url: string | undefined) {
    if (!url) {
      return null;
    }
    if (URL_PREFIX.test(url)) {
      return url;
    } else {
      if (this.getOption('baseUrl')) {
        return this.getOption('baseUrl') + url.replace(/\\/g, '/');
      }

      const relative = path.relative(
        path.dirname(this.activeLocation),
        path.dirname(url),
      );

      return path.join(relative, path.basename(url)).replace(/\\/g, '/');
    }
  }

  getBaseFrontmatterVars(page: PageEvent<Reflection>) {
    return {
      title: unEscapeChars(
        getReflectionTitle(page.model as DeclarationReflection),
      ),
    };
  }

  urlTo = (reflection: Reflection) => this.relativeURL(reflection.url);

  relativeURL = (url: string | undefined) => {
    return this.getRelativeUrl(url);
  };

  baseFrontmatterVars = (page: PageEvent<Reflection>) => {
    return this.getBaseFrontmatterVars(page);
  };

  templates = templates(this);
  partials = partials(this);
}
