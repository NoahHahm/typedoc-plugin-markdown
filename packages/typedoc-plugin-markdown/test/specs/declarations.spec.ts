import { ProjectReflection } from 'typedoc';
import { MarkdownThemeRenderContext } from '../../src/render-context';

describe(`Declarations:`, () => {
  let project: ProjectReflection;
  let context: MarkdownThemeRenderContext;

  beforeAll(async () => {
    ({ project, context } = await global.bootstrap(['declarations.ts'], {
      stubPartials: ['sources'],
    }));
  });

  test(`should compile a const with default value`, () => {
    expect(
      context.declarationMember(
        project.getChildByName('stringConstWithDefaultValue') as any,
        2,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile a let with default value`, () => {
    expect(
      context.declarationMember(
        project.getChildByName('stringLetWithDefaultValue') as any,
        2,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile an undefined declaration`, () => {
    expect(
      context.declarationMember(
        project.getChildByName('undefinedNumberDeclaration') as any,
        2,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile object literal declaration`, () => {
    expect(
      context.declarationMember(
        project.getChildByName('objectLiteralDeclaration') as any,
        2,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile object literal cast as a const`, () => {
    expect(
      context.declarationMember(
        project.getChildByName('objectLiteralAsConstDeclaration') as any,
        2,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile type literal declaration`, () => {
    expect(
      context.declarationMember(
        project.getChildByName('typeLiteralDeclaration') as any,
        2,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile declaration with double underscores in name and value`, () => {
    expect(
      context.declarationMember(
        project.getChildByName('__DOUBLE_UNDERSCORES_DECLARATION__') as any,
        2,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile any function type`, () => {
    expect(
      context.declarationMember(
        project.getChildByName('AnyFunctionType') as any,
        2,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile function declaration`, () => {
    expect(
      context.declarationMember(
        project.getChildByName('functionDeclaration') as any,
        2,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile indexable declaration`, () => {
    expect(
      context.declarationMember(
        project.getChildByName('indexableDeclaration') as any,
        2,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile enum declaration`, () => {
    expect(
      context.declarationMember(
        (project.getChildByName('EnumDeclarations') as any).children[0],
        2,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile enum declaration with defaults`, () => {
    expect(
      context.declarationMember(
        (project.getChildByName('EnumDeclarationsWithDefaults') as any)
          .children[0],
        2,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile declaration with accessors`, () => {
    expect(
      context.declarationMember(
        project.getChildByName('getterAndSetter') as any,
        2,
      ),
    ).toMatchSnapshot();
  });
});
