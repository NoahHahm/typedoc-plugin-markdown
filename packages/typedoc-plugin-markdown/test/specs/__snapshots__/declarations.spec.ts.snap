// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Declarations: should compile a const with default value 1`] = `
"> \`const\` **stringConstWithDefaultValue**: \`"hello"\` = \`'hello'\`

[partial: sources]

"
`;

exports[`Declarations: should compile a let with default value 1`] = `
"> \`let\` **stringLetWithDefaultValue**: \`string\` = \`'hello'\`

[partial: sources]

"
`;

exports[`Declarations: should compile an undefined declaration 1`] = `
"> \`let\` **undefinedNumberDeclaration**: \`number\`

[partial: sources]

"
`;

exports[`Declarations: should compile any function type 1`] = `
"> **AnyFunctionType**: <\`A\`> \`function\`

## Type parameters


| Parameter | Default |
| :------ | :------ |
| \`A\` | \`any\` |


## Parameters


| Parameter | Type |
| :------ | :------ |
| ...\`input\` | \`any\`[] |


## Returns

\`A\`



[partial: sources]

"
`;

exports[`Declarations: should compile declaration with accessors 1`] = `
"> \`let\` **getterAndSetter**: \`object\`

## Type declaration

### \`prop\`

> **prop**: \`string\`



***

### \`getMe\`

> \`get\` getMe(): [\`Promise\`]( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise )\\< \`string\` \\>



***

### \`setMe\`

> \`set\` setMe(\`x\`: \`string\`): \`void\`



[partial: sources]

"
`;

exports[`Declarations: should compile declaration with double underscores in name and value 1`] = `
"> \`const\` **\\_\\_DOUBLE\\_UNDERSCORES\\_DECLARATION\\_\\_**: *typeof* [\`__DOUBLE_UNDERSCORES_DECLARATION__\`](variables/variable._DOUBLE_UNDERSCORES_DECLARATION__.md)

[partial: sources]

"
`;

exports[`Declarations: should compile enum declaration 1`] = `
"> **Down**: \`1\`

[partial: sources]

"
`;

exports[`Declarations: should compile enum declaration with defaults 1`] = `
"> **East**: \`"East"\`

[partial: sources]

"
`;

exports[`Declarations: should compile function declaration 1`] = `
"> **functionDeclaration**

[partial: sources]

"
`;

exports[`Declarations: should compile indexable declaration 1`] = `
"> \`let\` **indexableDeclaration**: \`object\`

## Index signature

 \\[\`index\`: \`number\`\\]: \`string\`

## Type declaration

### \`arg1\`

> **arg1**: \`string\`



[partial: sources]

"
`;

exports[`Declarations: should compile object literal cast as a const 1`] = `
"> \`const\` **objectLiteralAsConstDeclaration**: \`object\`

## Type declaration

### \`Prop1\`

> \`readonly\` **Prop1**: \`"Prop1"\` = \`'Prop1'\`



***

### \`Prop2\`

> \`readonly\` **Prop2**: \`"Prop2"\` = \`'Prop2'\`



***

### \`Prop3\`

> \`readonly\` **Prop3**: \`"Prop3"\` = \`'Prop3'\`



[partial: sources]

"
`;

exports[`Declarations: should compile object literal declaration 1`] = `
"> \`const\` **objectLiteralDeclaration**: \`object\`

Comments for object literal

## Type declaration

### \`valueA\`

> **valueA**: \`number\` = \`100\`

Comment for valueA



***

### \`valueB\`

> **valueB**: \`boolean\` = \`true\`



***

### \`valueC\`

> **valueC**: \`object\` = \`{}\`



***

### \`valueX\`

> **valueX**: \`object\`

Comment for valueX

#### Type declaration

> ##### \`valueX.valueA\`
>
> > **valueA**: \`number\`[]
>
> Comment for valueA
>
> ##### \`valueX.valueB\`
>
> > **valueB**: \`object\`
>
> ###### Type declaration
>
> > ###### \`valueB.valueB1\`
> >
> > > **valueB1**: \`number\` = \`1\`
> >
> > ###### \`valueB.valueB2\`
> >
> > > **valueB2**: \`number\` = \`2\`
> >
> >
>
> ##### \`valueX.valueZ\`
>
> > **valueZ**: \`string\` = \`'foo'\`
>
> Comment for valueZ
>
>



***

### \`valueY\`

> **valueY**: \`function\`

Comment for value Y

#### Returns

\`string\`







***

### \`valueZ\`

> **valueZ**: \`string\` = \`'foo'\`

Comment for valueZ



[partial: sources]

"
`;

exports[`Declarations: should compile type literal declaration 1`] = `
"> \`let\` **typeLiteralDeclaration**: \`object\`

## Type declaration

### \`valueA\`

> \`optional\` **valueA**: \`number\`

Comments for valueA



***

### \`valueB\`

> \`optional\` **valueB**: \`boolean\`



***

### \`valueC\`

> **valueC**: \`object\`

#### Index signature

 \\[\`dataId\`: \`string\`\\]: \`"ok"\` \\| \`"ko"\`



***

### \`valueX\`

> **valueX**: \`object\`

Comment for valueX

#### Type declaration

> ##### \`valueX.valueA\`
>
> > **valueA**: \`number\`[]
>
> ##### \`valueX.valueY\`
>
> > **valueY**: \`function\`
>
> Nested comment for valueX.valueY
>
> ###### Parameters
>
>
> | Parameter | Type |
> | :------ | :------ |
> | \`z\` | \`string\` |
>
>
> ###### Returns
>
> ###### \`a\`
>
> > **a**: \`string\`
>
> ###### \`b\`
>
> > **b**: \`string\`
>
>
>
>
>
>
>
> ##### \`valueX.valueZ\`
>
> > **valueZ**: \`string\`
>
> Nested comment for valueX.valueZ
>
>



***

### \`valueY\`

> **valueY**: \`function\`

#### Returns

\`string\`





***

### \`valueZ\`

> **valueZ**: \`string\`

Comment for valueZ



[partial: sources]

"
`;
